lua/entities/kleiner/init.lua

```diff
@@ -0,0 +1,970 @@
+		local dmg = DamageInfo()
+		dmg:SetDamage(5)
+		dmg:SetAttacker(self)
+		self:TakeDamageInfo(dmg)
+	end
+	self.NextTargetTime = CurTime() + 2
+	self:SetTarget(nil)
+	return false
+end
+
+function ENT:ResetBehavior()
+	if(self.Suicidal)then return end
+	self.shifted = nil
+	if(self.path and IsValid(self.path))then 
+		self.path:Invalidate() 
+		self.path = nil 
+	end
+	self.ManualTarget = nil
+	self:FindTarget()
+end
+
+
+
+function ENT:IgnoreTarget(target)
+	self.TargetBlacklist = self.TargetBlacklist or {}
+	self.TargetBlacklist[target] = CurTime() + 1
+end
+
+function ENT:RunBehaviour()
+	-- This function is called when the entity is first spawned, it acts as a giant loop that will run as long as the NPC exists
+	while ( true ) do
+		
+		if ( IsValid(self:GetTarget()) ) then
+				self.loco:SetDesiredSpeed( (self.Suicidal and 350) or (self:GetBased() and 500) or 200 )		
+				self.loco:SetAcceleration(self.Suicidal and 400 or 200)
+				self.loco:SetDeceleration(self.Suicidal and 400 or 500)
+				self.loco:SetJumpHeight(100)
+				if(!self:NearTarget())then
+					local result = self:ChaseTarget( ) 
+					if(result == "failed")then self:IgnoreTarget(self:GetTarget())  self:ResetBehavior() end --if chasing the target fails somehow, its probably a wise assumption that its redundant to keep trying.
+				end
+				
+			
+			if(self:NearTarget())then
+				if(!self:IsTalking())then 
+					self:SayStuff() 	
+				end
+				self.loco:FaceTowards(self:GetTarget():GetPos())
+				if(!self.shifted)then
+					local result = self:MoveToPos( self:GetTarget():GetPos() + VectorRand()*Vector(1,1,0):GetNormalized()*math.Rand(80,150), {maxage=5} )
+					if(result != "ok")then self:ResetBehavior() end 
+					--cheap method of keeping kleiners with similar targets from clumping together. i figure it's cheaper than some kind of avoidance.
+					self.shifted = true
+				end
+				coroutine.wait(0.15)
+			else
+				self.shifted = nil
+			end
+			
+			if(self.ManualTarget == nil and math.random(1,100) == 1)then
+				self:ResetBehavior()
+			end
+	
+		else
+			-- no target, so we wander.
+			self.loco:SetDesiredSpeed( self.Suicidal and 350 or 200 )		
+			self.loco:SetAcceleration(self.Suicidal and 400 or 200)
+			self.loco:SetDeceleration(self.Suicidal and 900 or 900)
+			local wanderpos = self:FindSpot("random",{type="hiding",pos=self:GetPos(),radius=4000,stepup=900,stepdown=900})
+			if(wanderpos != nil)then
+			self:MoveToPos( wanderpos, {maxage=5} )
+			end
+			coroutine.wait(5)
+			if ( self.loco:IsStuck() ) then
+				self:HandleStuck()
+			end
+			if(math.random(1,5) == 1)then
+				self:ResetBehavior()
+			end
+			
+		end
+		-- At this point in the code the bot has stopped chasing the player or finished walking to a random spot
+		-- Using this next function we are going to wait 2 seconds until we go ahead and repeat it 
+		coroutine.wait(0.4)
+		
+	end
+
+end
+
+
+function ENT:MoveToPos( pos, options )
```

Comment:
wait, isn't this method built into the NextBot?

Reply:
i had to make an alternate implementation of it that runs  self:WhilePathing(), which allows movement to be interrupted for handling things like doors